### ISSUE TITLE
what is the right way to justify text?

### ISSUE THREAD

**Issue by AleXoundOS** [RyanGlScott/text-show#45]
There are [justifyLeft](http://hackage.haskell.org/package/text-1.2.3.1/docs/Data-Text.html#v:justifyLeft) and [justifyRight](http://hackage.haskell.org/package/text-1.2.3.1/docs/Data-Text.html#v:justifyRight) in the [Data.Text](http://hackage.haskell.org/package/text-1.2.3.1/docs/Data-Text.html) module of [text](http://hackage.haskell.org/package/text-1.2.3.1) package. Of course they are intended to work with values of [Text](http://hackage.haskell.org/package/text-1.2.3.1/docs/Data-Text.html#t:Text) type. Also, I've found some hints that their performance might be slow, since internally they use [replicateCharI](http://hackage.haskell.org/package/text-1.2.3.1/docs/src/Data.Text.Internal.Fusion.Common.html#replicateCharI), unsure.

Is it worth doing this instead?
```haskell
justifyLeftb :: Int -> Builder -> Builder
justifyLeftb w t =
  t <> mconcat (replicate (w - fromIntegral (lengthB t)) showbSpace)
```

Or should I resort on Data.Text module?

**Comment by RyanGlScott** [comment]
I've never used these functions before, so I can't offer any insights as to their performance. As is the case with every performance-oriented question, the only way to know for sure is to benchmark it.

**Comment by AleXoundOS** [comment]
If not taking into account performance characteristics, what is the idiomatic way to justify text when I use text-show package in your opinion?

Basically I need to arrange `showb` values in a table. There are many table-related packages out there, but none of them work with `Text`. They all use `String` which is considered harmful for such cases in real applications.

**Comment by RyanGlScott** [comment]
I don't know how to answer your question, if only because `text-show` wasn't designed with the use case of justifying text in mind. Especially since either of these `justify-*` functions can be implemented without needing to depend on `text-show` on all.

**Comment by AleXoundOS** [comment]
Using their `Text` implementation will force to escape `Builder`. As I understood while reading [documentation](http://hackage.haskell.org/package/text-show-3.7.5/docs/TextShow.html), that it is not good and it's better to stay within `Builder` until using the final `toText`. And that's why a couple of functions like `showbParen`, `showbCommaSpace`, etc exist.

Am I confused?

**Comment by RyanGlScott** [comment]
I think you might be confused as to the purpose of this library. `text-show` is intended to be a library for converting values to text values, not a general-purpose combinator library around the `text` library's `Builder` type. I only expose the `showbParen` and `showbCommaSpace` functions because code that `TextShow.TH` generates uses them directly, but I certainly never intended `text-show` to implement every possible `Builder`-manipulating function under the sun. Adding functions to do text justification feels like feature creep.

In short, it may be a good idea to implement your functions using `Builder`. Or it may not be, I'm not sure. But regardless of the answer to that question, I'm not really sure what the `text-show` library has to do with any of this.

**Comment by AleXoundOS** [comment]
Thank you. I've got your point. Also, I didn't realize that `Builder` is defined in the `text` package, since haddock doesn't show where the type comes from...
